File: .travis/bench-expectations.py
#!/usr/bin/env python3
"""Emit per-metric expected values for one (triple, device), for the bundle's
expectation-guided retry.

One `metric value` line per metric; value = recent median of the non-null points in
bench-data/<triple>/<device>.json. Keys are underscored, as stored in bench-data; the
bundle normalizes '-' -> '_' before looking a metric up. A device with no history
yields an empty file, which simply disables retry there (single-shot, as before).
"""
import argparse, json, os, statistics


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--bench-data", required=True)
    ap.add_argument("--triple", required=True)
    ap.add_argument("--device", required=True)
    ap.add_argument("--out", required=True)
    ap.add_argument("--window", type=int, default=10, help="trailing nights to median over")
    a = ap.parse_args()

    path = os.path.join(a.bench_data, a.triple, f"{a.device}.json")
    lines = []
    if os.path.exists(path):
        d = json.load(open(path))
        for m, arr in d.get("metrics", {}).items():
            vals = [v for v in arr[-a.window:] if v is not None]
            if vals:
                lines.append(f"{m} {statistics.median(vals)}")
    open(a.out, "w").write("".join(f"{l}\n" for l in lines))
    print(f"expectations: {len(lines)} metrics -> {a.out}")


if __name__ == "__main__":
    main()

File: cli/src/bench_expectations.rs
//! `tract bench-expectations`: emit per-metric `metric expected threshold` lines for
//! one (triple, device), so the suite retries exactly the benches whose value would
//! show as a PR red. Port of `.travis/bench-expectations.py`.
//!
//! `expected` is the recent median of the non-null points; `threshold` is the |Δ%|
//! that would make the metric a red (`bench_common::red_threshold`). Never-gated
//! metrics are omitted (not retried). A device with no history yields an empty file
//! (retry disabled there, single-shot).

use crate::bench_common::{Thresholds, median, red_threshold, series_noise};
use serde::Deserialize;
use std::collections::BTreeMap;
use tract_hir::internal::*;

#[derive(Deserialize)]
struct BenchData {
    #[serde(default)]
    metrics: BTreeMap<String, Vec<Option<f64>>>,
}

pub fn handle(matches: &clap::ArgMatches) -> TractResult<()> {
    let get = |k| matches.get_one::<String>(k).map(String::as_str);
    let bench_data = get("bench-data").context("--bench-data is required")?;
    let thresholds = get("thresholds").context("--thresholds is required")?;
    let triple = get("triple").context("--triple is required")?;
    let device = get("device").context("--device is required")?;
    let out = get("out").context("--out is required")?;
    let window: usize = get("window").map(str::parse).transpose()?.unwrap_or(10);

    let cfg = Thresholds::load(thresholds)?;
    let path = format!("{bench_data}/{triple}/{device}.json");

    let mut lines = vec![];
    if std::path::Path::new(&path).exists() {
        let data: BenchData = serde_json::from_str(&std::fs::read_to_string(&path)?)
            .with_context(|| format!("parsing {path}"))?;
        for (metric, arr) in &data.metrics {
            let vals: Vec<f64> =
                arr[arr.len().saturating_sub(window)..].iter().filter_map(|&v| v).collect();
            if vals.is_empty() {
                continue;
            }
            let expected = median(&vals);
            if let Some(thr) = red_threshold(metric, &cfg, series_noise(arr, 40, 8), Some(expected))
            {
                lines.push(format!("{metric} {expected} {thr}"));
            }
        }
    }

    std::fs::write(out, lines.iter().map(|l| format!("{l}\n")).collect::<String>())?;
    println!("expectations: {} gated metrics -> {out}", lines.len());
    Ok(())
}
